src/crimson/os/seastore/cache.cc

```diff
@@ -909,6 +900,17 @@ void Cache::complete_commit(
 {
   LOG_PREFIX(Cache::complete_commit);
   DEBUGT("enter", t);
+  auto& efforts = get_by_src(stats.committed_efforts_by_src,
+                             t.get_src());
+
+  // Transaction is now a go, set up in-memory cache state
+  // invalidate now invalid blocks
+  for (auto &i: t.retired_set) {
+    DEBUGT("retiring {}", t, *i);
+    get_by_ext(efforts.retire_by_ext,
+               i->get_type()).increment(i->get_length());
+    retire_extent(i);
+  }
```

Comment:
I only made some examples in the previous commit.

Seems to me `replace_extent(..)` has a similar issue as it is part of `Cache::prepare_record()` and it also makes transactional mutations visiable to others.

@athanatos what do you think? Should we move all the related logic that makes a committing transaction visible to `prepare_record()` or `complete_commit()`?

Reply:
> I only made some examples in the previous commit.
> 
> Seems to me `replace_extent(..)` has a similar issue as it is part of `Cache::prepare_record()` and it also makes transactional mutations visiable to others.

No, replace_extent doesn't have this issue, mutated extents has the io_promise protecting it


> 
> @athanatos what do you think? Should we move all the related logic that makes a committing transaction visible to `prepare_record()` or `complete_commit()`?

Reply:
IIUC there is an assumption that we need to resolve transaction conflicts at preparing phase synchronously, and any transaction entering the preparing phase must be valid to submit. So I think this implies we should commit the in-memory status only in `prepare_record()`, not in `complete_commit()`.

Reply:
> IIUC there is an assumption that we need to resolve transaction conflicts at preparing phase synchronously, and any transaction entering the preparing phase must be valid to submit. So I think this implies we should commit the in-memory status only in `prepare_record()`, not in `complete_commit()`.

Yes, and we can't simply move the `complete_commit `part into `prepare_record`, because new extents may have relative addresses which is only supposed to be seen by the same transaction. I think we probably need more design to fix this issue.

Reply:
One intuitive way to fix this, I think, is to add some kind of indicator, like `std::optional<seastar::promise> retiring_fut`, to `CachedExtent`, and we set it when rewriting the `CachedExtent` and resolve/unset it when the rewriting is finished; other transactions test and wait on this indicator, after which they might reread the extent.

Reply:
> new extents may have relative addresses which are only supposed to be seen by the transaction that created them.

Yeah, probably leveraging `CachedExtent::set_io_wait()` to also mark fresh extents as pending, and other transactions need to wait for the pending fresh extents until resolved and then add their correct paddrs into their transactions.

Reply:
> > new extents may have relative addresses which are only supposed to be seen by the transaction that created them.
> 
> Yeah, probably leveraging `CachedExtent::set_io_wait()` to also mark fresh extents as pending, and other transactions need to wait for the pending fresh extents until resolved and then add their correct paddrs into their transactions.

I think `CachedExtent::set_io_wait()` is semantically different, it is used to indicate the extents are being mutated, while, with regard to retiring, more operations need to be taken after the retiring is done compared to mutating extents, so we'd better add another indicator.

Reply:
> One intuitive way to fix this, I think, is to add some kind of indicator, like std::optional<seastar::promise> retiring_fut, to CachedExtent, and we set it when retiring the CachedExtent and resolve/unset it when the retiring is finished;

Hmm, maybe there is no need to add promise to the retiring extents? Other transactions only need to know that certain extents are retiring and just ignore them?

> other transactions test and wait on this indicator, after which they might reread the extent.

Why other transactions need to reread the extent that should be retired in the end?

Reply:
> > One intuitive way to fix this, I think, is to add some kind of indicator, like std::optionalseastar::promise retiring_fut, to CachedExtent, and we set it when retiring the CachedExtent and resolve/unset it when the retiring is finished;
> 
> Hmm, maybe there is no need to add promise to the retiring extents? Other transactions only need to know that certain extents are retiring and just ignore them?

they need to reference the new extent rather than the current old one.

> 
> > other transactions test and wait on this indicator, after which they might reread the extent.
> 
> Why other transactions need to reread the extent that should be retired in the end?

Sorry, by reread, I mean other transactions have to reference the new extent which probably need reiterating the corresponding onode/omap/lba tree.

Reply:
I think we can avoid the reiteration of the corresponding tree by adding a extent version link which links different versions of the same extent, is this feasible?

Reply:
I was intended to split the discussions for fresh extents and retired extents:
* fresh extents: I think `CachedExtent::set_io_wait()` could be used to mark fresh extents as not ready, just like io-wait for mutated extents. But I don't understand why it is semantically different and needs to add extra version links;
* retired extents: I think the only thing needed is to update the pinset at `prepare_record()` time, right?

Reply:
> I was intended to split the discussions for fresh extents and retired extents:
> 
> * fresh extents: I think `CachedExtent::set_io_wait()` could be used to mark fresh extents as not ready, just like io-wait for mutated extents; and also update the pinset at `prepare_record` time. But I don't understand why it is semantically different and needs to add extra version links;

I think fresh extents only need to be taken care of when they are associated with an retired extents. We don't need to care about pure fresh extents, they are not supposed to be seen by other transactions.

> * retired extents: I think the only thing needed is to update the pinset at `prepare_record()` time to make updates consistent, right?

I think the problem is a transaction tries to read an extent after it is retired and before the corresponding fresh extent is committed. I think this issue exists even if we update the pinset in `Cache::prepare_record()`.

Reply:
> when dealing with fresh extents associated with retired extents, more care needs to be taken after the fresh extent is committed than mutated extents, and that is we need to reference the fresh extent instead of the retired one AFTER the fresh extent is committed, we need a means to distinguish these kind of fresh extents from mutated extents.

I think you mean when the cleaner is rewriting extents, by moving an extent from the old paddr to a new paddr, causing a retired extent and an associated fresh extent. And when this transaction submits at `prepare_record` time, other transactions need to see the extent at the new paddr, right?

If so, that cleaner transaction should be able to update the in-memory lba tree consistently and synchronously at `prepare_record` time, so that when there is another transaction trying to read the laddr, it should be able to load a (probably-not-ready) extent with the new paddr (probably-still-relative) until it is ready. And any transactions already have the out-dated(moved) extents in the read-set should be invalidated and retry immediately.

> I think the problem is a transaction tries to read an extent after it is retired and before the corresponding fresh extent is committed.

Firstly, the same transaction should not read an extent after it is retired.

If it is not the same transaction, it should be invalidated if it's extent is modified(retired) by others. And if the extent is not yet in the read-set, it should be able to load a (probably-not-ready) extent with the new paddr (probably-still-relative) until it is ready. Is that correct?

Reply:
> I was intended to split the discussions for fresh extents and retired extents:
> 
> * fresh extents: I think `CachedExtent::set_io_wait()` could be used to mark fresh extents as not ready, just like io-wait for mutated extents; and also update the pinset at `prepare_record` time. But I don't understand why it is semantically different and needs to add extra version links;

`CachedExtent::set_io_wait()` works for pure fresh extents, but not fresh extents rewriting an retired one. Because, for those kind of extents, we need to make other transactions reference the fresh one after it is committed, which is different from mutated extents. This is what I meant by "semantically different". We need a means to distinguish these two kinds of extents ---- fresh extents rewriting an retired one and mutated extents.

I think for the 'reusing CachedExtent::set_io_wait()' approach to work, we need to do `add_extent` and update the pinset at `prepare_record` time,then we can use `CachedExtent::set_io_wait()` as at this moment there are no semantic difference between the above two kinds of extents. But at `prepare_record` time, the fresh extent's paddr might be relative and we need to correct it at `complete_commit` time, this would need to erase and re-add the extent to `Cache::extents` at the `complete_commit` time.

I think we can avoid this erase and re-add phase by keeping `add_extent` and the pinset update at `complete_commit` time, and instead, we add a mechinary like `CachedExtent::io_wait_promise` to make other transactions wait for the rewriting of extents.

Specifically, I think we can 1) re-enable the `extent_state_t::RETIRED`; 2) we invalidate conflicting transactions and set the retired extents to `extent_state_t::RETIRED` at `prepare_record` time; 3) we do the actual `retire_extent` work at `complete_commit` time; 4) when trying to get an extent, we check and the extent's state, if it is `extent_state_t::RETIRED`, we go referencing the corresponding fresh extent after the `CachedExtent::io_wait_promise` is resolved.

We can reference the corresponding fresh extent by re-searching the corresponding trees, which might be a little bit cpu-consuming. So I think maybe we can an extra version link, and follow the link to reference the fresh extent.

> * retired extents: I think the only thing needed is to update the pinset at `prepare_record()` time to make updates consistent, right?

Reply:
> If it is not the same transaction, it should be invalidated if it's extent is modified(retired) by others. And if the extent is not yet in the read-set, it should be able to load a (probably-not-ready) extent with the new paddr (probably-still-relative) until it is ready. Is that correct?

This is where we need to involve the extent erasing from and re-adding to `Cache::extents` at `complete_commit` time.

Reply:
> This is where we need to involve the extent erasing from and re-adding to Cache::extents at complete_commit time to change the relative addr to an absolute one.

I think I understand your point now. This is more tricky than I thought, and this also means relative addresses are no longer transaction-local when exposed to `Cache`, and we need to find a way to distinguish relative paddrs from different transactions when they are submitting.

Reply:
> I think I understand your point now. This is more tricky than I thought, and this also means relative addresses are no longer transaction-local if exposed to `Cache`, and we will need to find a way to distinguish relative paddrs from different transactions when they are submitting.

I think we can avoid exposing relative addresses by the method I mentioned [here](https://github.com/ceph/ceph/pull/43060#discussion_r703205913), what do you think about it?

Reply:
Emm, I cannot think of a case that it won't work. Maybe it is better named `extent_state_t::REWRITING` to be more explicit.

Reply:
yes, that's better